File: leetcode2/2medium/최원준/Q117.java
package Leetcode.최원준;

/*
1. 아이디어 :


2. 시간복잡도 :
O( )

3. 자료구조/알고리즘 :

 */

public class Q117 {
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};


    class Solution {
        Map<Integer, Node> depthsMap = new HashMap<>();
        public void postOrder(Node node, int depths) {
            if (node == null) return;

            postOrder(node.right, depths+1);
            postOrder(node.left, depths+1);

            node.next = depthsMap.getOrDefault(depths, null);
            depthsMap.put(depths, node);
        }

        public Node connect(Node root) {
            postOrder(root, 0);
            return root;
        }
    }
 */
}

File: leetcode2/2medium/강병호/Q117.py
class Leetcode.이재훈.이재훈.Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return root
        
        current = root
        temp = Node(0) 
        tail = temp
        
        while current:
            if current.left:
                tail.next = current.left
                tail = tail.next
            if current.right:
                tail.next = current.right
                tail = tail.next
            current = current.next
            
            if not current:
                current = temp.next 
                temp.next = None 
                tail = temp
        
        return root